### ISSUE TITLE
301 Redirect from /page to /page/

### ISSUE THREAD

**Issue by adampetrie** [laurilehmijoki/s3_website#233]
Im not sure if I've misunderstood something or not but I am attempting to use the redirect directive to create a 301 redirect from the `/content` version of a url to the `/content/` version that s3 will serve when it finds the resource at `/content/index.html`.

I know that by default s3 will serve a 302 and I'd like to make it a 301 for SEO purposes. Here's a snippet from my config:

```
redirects:
  /content: /content/
```

When I push the site the output says that the redirect is created but I cannot find anything confirming the redirect and when I curl the original URL the response is still a 302 redirect.

Am I missing something or perhaps is it not possible to change the default 302 to a 301?

Any input is helpful. Thanks!

**Comment by derekperkins** [comment]
I've been trying to figure out how to do that and haven't been able to figure it out.

**Comment by vitaly** [comment]
This is definitely supported by amazon: http://docs.aws.amazon.com/AmazonS3/latest/dev/IndexDocumentSupport.html

when you access /foo it first looks for an object /foo, and only if its not found does it look for /foo/index.html, which if found produces a 302 redirect /foo -> /foo/

**Comment by jhabdas** [comment]
Great! Can this be closed?

**Comment by iloveip** [comment]
Hi @adampetrie,
Have you figured out how to make 301 redirects? I have the same problem. But as the docs say, the root forward slash should be omitted in the requested page. Also, I think, it should end with `.html`:
```
redirects:
  content.html: /content/
```

**Comment by jhabdas** [comment]
@iloveip did you see @vitaly comment?

**Comment by iloveip** [comment]
@jhabdas Yes, sure. But he is talking about 302 redirects, and I need to make 301 redirects for SEO purposes.

**Comment by jhabdas** [comment]
Did you use `rel=canonical`? If so you should be fine. But don't take my word for it. Test.

**Comment by iloveip** [comment]
I'm trying to make redirects through `s3_website.yml` file, instead of creating new pages. As far as I understand, it is possible, but for some reason it doesn't seem to work. I created a new issue #280 with my settings.

**Comment by jhabdas** [comment]
If you're hosting on S3 the redirect happens automatically for existing pages. So if you're using canonicalization you should be good. But the only way to know for sure is to A/B test. If you can come back with some solid metrics this issue might get some traction. Otherwise it's really up to the crawler, and crawlers know AWS does this (or at least I would hope so).

**Comment by jhabdas** [comment]
Hope that makes sense. It's a stumbling block for sure. I hit it. I've seen others hit it. And so I'm just telling you what I know. And I know DMIAC is the only way to be certain.

**Comment by iloveip** [comment]
How can I use canonicalization with redirects set in `s3_website.yml` file? Should I create these pages manually to set `rel=canonical`? If so, how can I add the proper metadata to them?

**Comment by jhabdas** [comment]
Just a sec. I'm digging. To validate my understanding, is it true when I say you're attempting to use the `redirects` object in the s3 config file to perform site-side redirects for pretty URLs to ensure they have a trailing `/` (disregarding how the redirect happens)?

**Comment by jhabdas** [comment]
Yep. Just re-reviewed the docs I referenced above (for folders) and looked a level up to see if anything for static websites has changed as AWS recently with regard to how the folder -> folder/index.html redirects occur and it's still a 302.

So if my assumption above is correct, and you're attempting to use folders with an `index.html` sitting in each one and achieve a 301 you're in the right place, and AWS doesn't handle that (therefore this lib couldn't either as far as I can be certain).

> How can I use canonicalization with redirects set in `s3_website.yml` file?

You don't. You use the redirects in the s3 config for redirects and live with the 302 for folders redirecting to a nested `index.html` file. Your alternatives would be:

- Configure your SSG to use ugly URLs
- Go on a witch hunt at Amazon headquarters

If you add the canonicalization, bots will know where your pages live and _should not_ penalize you. That said, you'll still want to ensure your internal link structure points to the `/` version of the links to prevent an unnecessary redirect.

Beyond that, and I'm curious too, if you want to know what a 301 would do you could try A/B testing once your organic traffic stabilizes (if it does) and try measuring 301 using some other means.

**Comment by iloveip** [comment]
So do you mean that AWS can't handle redirects from `page.html` to `/page/`, because `/page/` has an `index.html` file in it and it already uses a 302 redirect?

**Comment by jhabdas** [comment]
Interesting. So it looks like you've already started using ugly URLs and are attempting to use the redirects object to prettify them. As far as I know the object doesn't support globbing for wildcards so if you're shooting to do this for more than a handful of pages it's going to hurt your deployment speed (redirects seem to be the slowest part of the deployment process).

But since your use case is a little different from the origin case mentioned when this post was opened I'm not entirely certain - but if you're having issue with it my gut tells me it's because of the special way S3 handles folders. But guts can be wrong.

I use redirects on my site to hide certain URL shaping I'm doing, e.g.

```yaml
redirects:
  gear: /gear/southeast-asia-carry-on-packing-list/`
```

That works fine. But it doesn't have a file extension. So if your scenario _with_ an extension isn't working - and it works with the regular AWS CLI, you've either nailed a bug or found an enhancement for this library which is just a hair different than this particular issue.

Hope that makes sense. 👍 👎

**Comment by iloveip** [comment]
Actually, I don't know, if I need an extension for the requested page at all. For example, I have only one page `/kb/page/`. But the search engine some how indexed this page as `/kb/page`, and now it tells that `/kb/page/` is a double. So I want to create a 301 redirect for this page to point to `/kb/page/`.

**Comment by iloveip** [comment]
@jhabdas Thank you very much for your help! I've deleted `.html` extensions for requested pages and now everything works :)

**Comment by laurilehmijoki** [comment]
@iloveip glad to hear that your problem was solved! Please close this issue if you consider this case closed.

**Comment by jhabdas** [comment]
@iloveip pleasure to be at your service. @laurilehmijoki her issue was a little different than the OP. that said this can be closed as in my mind it's functioning as expected.

**Comment by iloveip** [comment]
@laurilehmijoki I'm not the author, so I can't close it)

**Comment by jhabdas** [comment]
@iloveip you might find this interesting: https://moz.com/blog/meta-referrer-tag. the suggestion is even a 301 is thought to cause about a 15% loss in equity of a link. haven't seen any research on what a 302 might do, but if your URLs are well structured they'll always end in a `/` and anytime someone copies or bookmarks them they'll get the whole link (unless they're using their mouse, in which case they should be shot).

**Comment by iloveip** [comment]
Hi @jhabdas 

Thank you very much for the link! We are right on the point of moving our site from http to https.

**Comment by nabilfreeman** [comment]
Hi guys - I also encountered this problem when moving my site over to Jekyll.

We already had loads of indexed pages without trailing slashes, so I used AWS Lambda@Edge (linked to CloudFront) to rewrite the URLs to have a trailing slash - WITH A 301 CODE! 🎉

It's quite complicated, but this article is pretty good to start: https://read.acloud.guru/supercharging-a-static-site-with-lambda-edge-da5a1314238b

**Comment by jhabdas** [comment]
@nabilfreeman thanks for sharing. please keep us informed of any changes in organic traffic you do or don't receive as a result of this chance so we know whether or not it's worth the trouble (regardless of process used)
